providers/implementations/keymgmt/ml_kem_kmgmt.c

```diff
@@ -0,0 +1,529 @@
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_PRIV_KEY, NULL, 0),
+        OSSL_PARAM_END
+    };
+
+    return arr;
+}
+
+/*
+ * It is assumed the key is guaranteed non-NULL here, and is from this provider
+ */
+static int ml_kem_get_params(void *vkey, OSSL_PARAM params[])
+{
+    ML_KEM_KEY *key = vkey;
+    const ML_KEM_VINFO *v = ossl_ml_kem_key_vinfo(key);
+    OSSL_PARAM *p;
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_BITS);
+    if (p != NULL)
+        if (!OSSL_PARAM_set_int(p, v->bits))
+            return 0;
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_SECURITY_BITS);
+    if (p != NULL)
+        if (!OSSL_PARAM_set_int(p, v->secbits))
+            return 0;
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_MAX_SIZE);
+    if (p != NULL)
+        if (!OSSL_PARAM_set_int(p, v->ctext_bytes))
+            return 0;
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY);
+    if (p != NULL && ossl_ml_kem_have_pubkey(key)) {
+        if (p->data_type != OSSL_PARAM_OCTET_STRING)
+            return 0;
+        p->return_size = v->pubkey_bytes;
+        if (p->data != NULL) {
+            if (p->data_size < p->return_size)
+                return 0;
+            if (!ossl_ml_kem_encode_public_key(p->data, p->return_size, key))
+                return 0;
+        }
+    }
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_PRIV_KEY);
+    if (p != NULL && ossl_ml_kem_have_prvkey(key)) {
+        if (p->data_type != OSSL_PARAM_OCTET_STRING)
+            return 0;
+        p->return_size = v->prvkey_bytes;
+        if (p->data != NULL) {
+            if (p->data_size < p->return_size)
+                return 0;
+            if (!ossl_ml_kem_encode_private_key(p->data, p->return_size, key))
+                return 0;
+        }
+    }
+    return 1;
+}
+
+static const OSSL_PARAM *ml_kem_settable_params(void *provctx)
+{
+    static const OSSL_PARAM arr[] = {
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY, NULL, 0),
+        OSSL_PARAM_END
+    };
+
+    return arr;
+}
+
+static int ml_kem_set_params(void *vkey, const OSSL_PARAM params[])
+{
+    ML_KEM_KEY *key = vkey;
+    const ML_KEM_VINFO *v = ossl_ml_kem_key_vinfo(key);
+    const OSSL_PARAM *p;
+    const void *pubenc = NULL;
+    size_t publen = 0;
+
+    if (ossl_param_is_empty(params))
+        return 1;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY);
+    if (p == NULL)
+        return 1;
+
+    /* Key mutation is reportedly generally not allowed */
+    if (ossl_ml_kem_have_pubkey(key)) {
+        ERR_raise_data(ERR_LIB_PROV,
+                       PROV_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE,
+                       "ML-KEM keys cannot be mutated");
+        return 0;
+    }
+    /* An unlikely failure mode is the parameter having some unexpected type */
+    if (!OSSL_PARAM_get_octet_string_ptr(p, &pubenc, &publen))
```

Comment:
No, I think you're confused. The parameter is still a regular octet string, this is just a non-copying accessor for octet strings, that returns the address of the data inside the parameter. This interface works correctly, demonstrating as much.

The value is not retained by this function, so it need not be copied.

Reply:
Relevant code: https://github.com/openssl/openssl/blob/e0ea913f11cf64d000556bbf7cb9f8acdf6be4cb/crypto/params.c#L1676-L1687
https://github.com/openssl/openssl/blob/e0ea913f11cf64d000556bbf7cb9f8acdf6be4cb/crypto/params.c#L1647-L1662
Checks that the parameter type is an octet-string and returns a *pointer* to, rather than a copy of, the data.